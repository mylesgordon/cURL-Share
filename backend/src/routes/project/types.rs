use actix_session::SessionGetError;
use actix_web::HttpResponse;
use serde::{Deserialize, Serialize};

use crate::routes::types::UserError;

#[derive(Debug)]
pub enum ProjectError {
    Forbidden(String),
    ProjectDoesNotExistError(String),
    SessionGetError(String),
    SqlxError(sqlx::Error),
    UserError(UserError),
}

impl From<SessionGetError> for ProjectError {
    fn from(e: SessionGetError) -> Self {
        match e {
            // TODO: improve
            _ => ProjectError::SessionGetError(e.to_string()),
        }
    }
}

impl From<sqlx::Error> for ProjectError {
    fn from(e: sqlx::Error) -> Self {
        match e {
            sqlx::Error::RowNotFound => {
                ProjectError::ProjectDoesNotExistError("Project was not found".to_string())
            }
            _ => ProjectError::SqlxError(e),
        }
    }
}

impl From<UserError> for ProjectError {
    fn from(e: UserError) -> Self {
        match e {
            // TODO: improve
            UserError::SessionGetError(e) => ProjectError::SessionGetError(e),
            _ => ProjectError::UserError(e),
        }
    }
}

impl From<ProjectError> for HttpResponse {
    fn from(e: ProjectError) -> Self {
        match e {
            ProjectError::ProjectDoesNotExistError(_) => HttpResponse::NotFound().finish(),
            ProjectError::SessionGetError(_) => HttpResponse::Unauthorized().finish(),
            ProjectError::Forbidden(_) => HttpResponse::Forbidden().finish(),
            _ => HttpResponse::InternalServerError().finish(),
        }
    }
}

#[derive(Serialize)]
pub struct ProjectId {
    pub id: i64,
}

#[derive(Debug, Default, Deserialize)]
pub struct ProjectParams {
    pub search: String,
    pub labels: Vec<String>,
}
