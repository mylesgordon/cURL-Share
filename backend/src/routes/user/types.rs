use actix_web::HttpResponse;
use secrecy::Secret;
use serde::Deserialize;

#[derive(Debug)]
pub enum UserError {
    PasswordHashError(argon2::password_hash::Error),
    SqlxDatabaseError,
    SqlxError(sqlx::Error),
    UserAlreadyExists,
    UserNotFound(sqlx::Error),
}

impl From<argon2::password_hash::Error> for UserError {
    fn from(e: argon2::password_hash::Error) -> Self {
        Self::PasswordHashError(e)
    }
}

impl From<sqlx::Error> for UserError {
    fn from(e: sqlx::Error) -> Self {
        match e {
            sqlx::Error::RowNotFound => Self::UserNotFound(e),
            sqlx::Error::Database(database_err) => {
                if let Some(database_err_code) = database_err.code() {
                    match database_err_code.as_ref() {
                        "2067" => Self::UserAlreadyExists,
                        _ => Self::SqlxDatabaseError,
                    }
                } else {
                    Self::SqlxDatabaseError
                }
            }
            _ => Self::SqlxError(e),
        }
    }
}

impl From<UserError> for HttpResponse {
    fn from(e: UserError) -> Self {
        match e {
            UserError::UserAlreadyExists => HttpResponse::Conflict().into(),
            UserError::UserNotFound(_) => HttpResponse::Unauthorized().into(),
            _ => HttpResponse::InternalServerError().into(),
        }
    }
}

#[derive(Deserialize)]
pub struct UserRequest {
    pub username: String,
    pub password: Secret<String>,
}
