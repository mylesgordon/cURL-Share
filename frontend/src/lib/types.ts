export type Fetch = typeof fetch;

export enum ButtonType {
	Up = -1,
	Down = 1
}

export type CreateProjectResponse = {
	id: number;
};

export type Curl = {
	id: number;
	name: string;
	description: string;
	rawQuery: string;
};

export type CurlGroup = {
	id: number;
	curls: string;
	description: string;
	labels: string;
	name: string;
	project_id: number;
};

export type Project = {
	info: ProjectInfo;
	admins: Array<string>;
	collaborators: Array<string>;
	groups: Array<CurlGroup>;
};

export type ProjectInfo = {
	id: number;
	environments: string;
	description: string;
	name: string;
	visibility: string;
};

export enum Visibility {
	Public = 'Public',
	Private = 'Private'
}

export const VisibilityOptions = [
	{
		name: 'visibility',
		value: 'Public',
		label: 'Public'
	},
	{
		name: 'visibility',
		value: 'Private',
		label: 'Private'
	}
];
