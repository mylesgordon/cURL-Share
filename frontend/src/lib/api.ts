import { backendUrl } from './stores';
import { get } from 'svelte/store';
import type { Project, ProjectInfo } from './types';

const url = get(backendUrl);

type CreateProjectResponse = {
	id: number;
};

type ProjectAdminStatus = {
	isUserAdmin: boolean;
}

type Fetch = typeof fetch;

export async function fetchIsLoggedIn(fetch: Fetch): Promise<boolean> {
	const response = await fetch(`${url}/api/v1/user-status`, {
		method: 'GET',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});
	const responseJson = await response.json();
	return responseJson.is_logged_in ? responseJson.is_logged_in : false;
}

export async function fetchProject(fetch: Fetch, projectId: string): Promise<Project> {
	const projectResponse = await fetch(`${url}/api/v1/project/${projectId}`, {
		method: 'GET',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});

	return await projectResponse.json();
}

export async function fetchProjectAdminStatus(fetch: Fetch, projectId: string): Promise<ProjectAdminStatus> {
	const projectResponse = await fetch(`${url}/api/v1/project/${projectId}/is-user-admin`, {
		method: 'GET',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});

	return await projectResponse.json();
}

export async function fetchProjects(fetch: Fetch): Promise<ProjectInfo[]> {
	const response = await fetch(`${url}/api/v1/project`, {
		method: 'GET',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});
	return await response.json();
}

export async function createProjectRequest(
	fetch: Fetch,
	projectInfo: ProjectInfo
): Promise<CreateProjectResponse> {
	const response = await fetch(`${url}/api/v1/project`, {
		method: 'POST',
		body: JSON.stringify(projectInfo),
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url, 'Content-Type': 'application/json' },
		credentials: 'include'
	});

	return await response.json();
}

export async function deleteProjectRequest(fetch: Fetch, project_id: number): Promise<void> {
	await fetch(`${url}/api/v1/project/${project_id}`, {
		method: 'DELETE',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});
}

export async function updateProjectRequest(fetch: Fetch, project: Project): Promise<void> {
	await fetch(`${url}/api/v1/project/${project.info.id}`, {
		method: 'POST',
		body: JSON.stringify(project),
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url, 'Content-Type': 'application/json' },
		credentials: 'include'
	});
}

export async function logInRequest(
	fetch: Fetch,
	endpoint: string,
	username: string,
	password: string
): Promise<number> {
	const request = await fetch(`${url}/api/v1/${endpoint}`, {
		method: 'POST',
		mode: 'cors',
		headers: {
			'Access-Control-Allow-Origin': url,
			'Content-Type': 'application/json'
		},
		body: JSON.stringify({ username, password }),
		credentials: 'include'
	});

	return request.status;
}

export async function logOutRequest(fetch: Fetch): Promise<void> {
	await fetch(`${url}/api/v1/log-out`, {
		method: 'POST',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});
}
