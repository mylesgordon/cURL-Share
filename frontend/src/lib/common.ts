export function splitAndTrim(input: string): Array<string> {
		return input.split(',').map((item) => item.trim());
	}