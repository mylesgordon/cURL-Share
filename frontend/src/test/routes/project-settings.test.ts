import '@testing-library/jest-dom';
import { type Project, Visibility } from '$lib/types';
import { deleteProjectRequest, updateProjectRequest } from '$lib/api.page';
import { fireEvent, render, screen } from '@testing-library/svelte';
import { goto } from '$app/navigation';
import { testMatchingSnapshot } from '../common';
import ProjectSettingsPage from '../../routes/project/[slug]/settings/+page.svelte';
import userEvent from '@testing-library/user-event';
import type { ComponentProps } from 'svelte';

vi.mock('$lib/api.page');
vi.mock('$app/navigation');

describe('Project page', () => {
	type ProjectSettingsPageProps = ComponentProps<ProjectSettingsPage>;
	const props: ProjectSettingsPageProps = {
		data: {
			isLoggedIn: true,
			project: {
				admins: ['admin1', 'admin2'],
				collaborators: ['collaborator1', 'dave'],
				groups: [],
				info: {
					id: 1,
					name: 'Placeholder project',
					description: 'This is a placeholder project',
					visibility: Visibility.Public,
					environments: 'https://cool.com, https://dev.cool.com'
				}
			},
			isUnitTest: true,
			success: true,
			userAdminStatus: { isUserAdmin: true }
		}
	};
	const propsNotAdmin = JSON.parse(JSON.stringify(props));
	propsNotAdmin.data.userAdminStatus.isUserAdmin = false;

	const propsSuccessFalse = JSON.parse(JSON.stringify(props));
	propsSuccessFalse.data.success = false;

	testMatchingSnapshot(ProjectSettingsPage, props);
	testMatchingSnapshot(ProjectSettingsPage, propsNotAdmin);
	testMatchingSnapshot(ProjectSettingsPage, propsSuccessFalse);

	it('should display the correct title', () => {
		render(ProjectSettingsPage, props);
		expect(document.title).toEqual('Placeholder project settings | cURL Share');
	});

	it("should display the delete dialog and disappear when 'no' is clicked", async () => {
		render(ProjectSettingsPage, props);
		const dialog = screen.getByRole('dialog', { hidden: true });
		expect(dialog).toHaveAttribute('aria-hidden', 'true');

		const deleteButton: HTMLInputElement = screen.getByText('Delete');
		await fireEvent.click(deleteButton);

		const noButton = screen.getByText('No');
		await fireEvent.click(noButton);
		expect(dialog).toHaveAttribute('aria-hidden', 'true');
	});

	it("should send a delete request and redirect the user when 'yes' is clicked on the delete dialog", async () => {
		render(ProjectSettingsPage, props);
		const deleteButton: HTMLInputElement = screen.getByText('Delete');
		await fireEvent.click(deleteButton);

		const yesButton = screen.getByText('Yes');
		await fireEvent.click(yesButton);

		expect(deleteProjectRequest).toHaveBeenCalledWith(window.fetch, 1);
		expect(deleteProjectRequest).toHaveBeenCalledTimes(1);

		expect(goto).toHaveBeenCalledWith('/');
		expect(goto).toHaveBeenCalledTimes(1);
	});

	it('should send a correct POST request when the user updates project settings', async () => {
		render(ProjectSettingsPage, props);

		const newName = 'New Name';
		const newDescription = 'New description';

		const nameInput = screen.getByLabelText('Name');
		const descriptionInput = screen.getByLabelText('Description');
		const saveButton = screen.getByText('Save');

		await userEvent.clear(nameInput);
		await userEvent.clear(descriptionInput);

		await userEvent.type(nameInput, newName);
		await userEvent.type(descriptionInput, newDescription);
		await userEvent.click(saveButton);

		const expectedProjectData: Project = { ...props.data.project };
		expectedProjectData.info.name = newName;
		expectedProjectData.info.description = newDescription;

		expect(updateProjectRequest).toHaveBeenCalledWith(window.fetch, expectedProjectData);
		expect(updateProjectRequest).toHaveBeenCalledTimes(1);

		expect(goto).toHaveBeenCalledWith('/project/1');
	});
});
