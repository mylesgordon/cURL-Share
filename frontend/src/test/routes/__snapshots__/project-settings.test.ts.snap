// Vitest Snapshot v1

exports[`Project page > should match snapshot 1`] = `
<body>
  <div>
    <!--&lt;Meta&gt;-->
     
    <h1>
      Placeholder project settings
    </h1>
     
    <form
      class="flex flex-col space-y-3 pt-4"
    >
      <div
        class="w-100 s-JZthjkXlZrq3"
      >
        <label
          class="label s-JZthjkXlZrq3"
          for="name"
        >
          Name
        </label>
         
        <input
          class="input input-rounded s-JZthjkXlZrq3"
          id="name"
          type="text"
        />
         
      </div>
      <!--&lt;Input&gt;-->
       
      <div
        class="w-100 s-JZthjkXlZrq3"
      >
        <label
          class="label s-JZthjkXlZrq3"
          for="description"
        >
          Description
        </label>
         
        <input
          class="input input-rounded s-JZthjkXlZrq3"
          id="description"
          type="text"
        />
         
      </div>
      <!--&lt;Input&gt;-->
       
      <fieldset
        class="radio-group s-K6jzjY3vjtlg"
      >
        <legend
          class="radio-legend s-K6jzjY3vjtlg"
        >
          Visibility
        </legend>
         
        <label
          class="radio-label-wrap s-K6jzjY3vjtlg"
        >
          <input
            class="radio s-K6jzjY3vjtlg"
            id="visibility-visibility-0"
            name="visibility"
            type="radio"
            value="Public"
          />
           
          <span
            aria-hidden="true"
            class="radio-label s-K6jzjY3vjtlg"
          />
           
          <span
            class="radio-label-copy s-K6jzjY3vjtlg"
          >
            Public
          </span>
           
        </label>
        <label
          class="radio-label-wrap s-K6jzjY3vjtlg"
        >
          <input
            class="radio s-K6jzjY3vjtlg"
            id="visibility-visibility-1"
            name="visibility"
            type="radio"
            value="Private"
          />
           
          <span
            aria-hidden="true"
            class="radio-label s-K6jzjY3vjtlg"
          />
           
          <span
            class="radio-label-copy s-K6jzjY3vjtlg"
          >
            Private
          </span>
           
        </label>
      </fieldset>
      <!--&lt;ChoiceInput&gt;-->
      <!--&lt;VisibilityInput&gt;-->
       
      <div
        class="w-100 s-JZthjkXlZrq3"
      >
        <label
          class="label s-JZthjkXlZrq3"
          for="admins"
        >
          Administrators (comma separated)
        </label>
         
        <input
          class="input input-rounded s-JZthjkXlZrq3"
          id="admins"
          type="text"
        />
         
      </div>
      <!--&lt;Input&gt;-->
       
      <div
        class="w-100 s-JZthjkXlZrq3"
      >
        <label
          class="label s-JZthjkXlZrq3"
          for="collaborators"
        >
          Collaborators (comma separated)
        </label>
         
        <input
          class="input input-rounded s-JZthjkXlZrq3"
          id="collaborators"
          type="text"
        />
         
      </div>
      <!--&lt;Input&gt;-->
       
      <div
        class="w-100 s-JZthjkXlZrq3"
      >
        <label
          class="label s-JZthjkXlZrq3"
          for="environments"
        >
          Environments (comma separated)
        </label>
         
        <input
          class="input input-rounded s-JZthjkXlZrq3"
          id="environments"
          type="text"
        />
         
      </div>
      <!--&lt;Input&gt;-->
       
      <div
        class="flex flex-row gap-2 justify-center"
      >
        <button
          class="btn btn-primary btn-bordered btn-rounded s-NAQuwh5ehqbV"
          type="submit"
        >
          Save
        </button>
        
        <!--&lt;Button&gt;-->
         
        <button
          class="btn btn-secondary btn-bordered btn-rounded s-NAQuwh5ehqbV"
          type="button"
        >
          Delete
        </button>
        
        <!--&lt;Button&gt;-->
      </div>
    </form>
     
    
    <!--&lt;SvelteA11yDialog&gt;-->
    <!--&lt;Dialog&gt;-->
     
    <div
      id="dialog-root"
    >
      <div
        aria-hidden="true"
        aria-labelledby="deleteProjectDialog"
        class="dialog"
        id="a11y-dialog"
        role="dialog"
      >
        <div
          class="dialog-overlay"
          data-a11y-dialog-hide=""
          tabindex="-1"
        />
         
        <div
          class="dialog-content dialog-fade-in"
          role="document"
        >
           
          <p
            class="h4 mbe18 flex justify-center"
            id="deleteProjectDialog"
          >
            Delete project
          </p>
           
          <div
            class="flex flex-col items-center gap-2"
          >
            <h2>
              Are you sure you want to delete 
              Placeholder project
              ?
            </h2>
             
            <form>
              <button
                class="btn btn-secondary btn-bordered btn-rounded s-NAQuwh5ehqbV"
                type="submit"
              >
                Yes
              </button>
              
              <!--&lt;Button&gt;-->
               
              <button
                class="btn btn-primary btn-bordered btn-rounded s-NAQuwh5ehqbV"
                type="button"
              >
                No
              </button>
              
              <!--&lt;Button&gt;-->
            </form>
          </div>
           
          <button
            aria-label="Close delete dialog"
            class="dialog-close dialog-close-button"
            data-a11y-dialog-hide=""
            type="button"
          >
            <div
              class="close-button close-button-large s-C9mYFMcjNvxm"
            >
              <svg
                aria-hidden="true"
                class="close s-C9mYFMcjNvxm"
                viewBox="0 0 24 24"
              >
                <path
                  class="s-C9mYFMcjNvxm"
                  d="M.439 21.44a1.5 1.5 0 0 0 2.122 2.121l9.262-9.261a.25.25 0 0 1 .354 0l9.262 9.263a1.5 1.5 0 1 0 2.122-2.121L14.3 12.177a.25.25 0 0 1 0-.354l9.263-9.262A1.5 1.5 0 0 0 21.439.44L12.177 9.7a.25.25 0 0 1-.354 0L2.561.44A1.5 1.5 0 0 0 .439 2.561L9.7 11.823a.25.25 0 0 1 0 .354Z"
                  fill="currentColor"
                />
              </svg>
            </div>
            
            <!--&lt;Close&gt;-->
          </button>
        </div>
      </div>
    </div>
    
    <!--&lt;+page&gt;-->
  </div>
</body>
`;

exports[`Project page > should match snapshot 2`] = `
<body>
  <div>
    <p>
      Failed to load project settings
    </p>
     
    <div
      id="dialog-root"
    />
    
    <!--&lt;+page&gt;-->
  </div>
</body>
`;

exports[`Project page > should match snapshot 3`] = `
<body>
  <div>
    <p>
      Failed to load project settings
    </p>
     
    <div
      id="dialog-root"
    />
    
    <!--&lt;+page&gt;-->
  </div>
</body>
`;
